title: Логирование стека ошибок

(скрипт: DB/Example/stack-error-log.sql)

Пример использования №1:

(code)

declare
  lg lg_logger_t := lg_logger_t.getLogger(
    moduleName => 'Test'
    , objectName => 'TestBlock'
  );

  procedure internal
  is
  begin
    raise_application_error(
      pkg_Error.ProcessError
      , 'Произошла ошибка' || lpad( '!', 1000, '_')
    );
  exception when others then
    raise_application_error(
      pkg_Error.ErrorStackInfo
      , lg.errorStack( 'Ошибка "Internal"' || lpad( '!', 1000, '_'))
      , true
    );
  end internal;

begin
  internal();
exception when others then
  pkg_Common.outputMessage(
    lg.getErrorStack()
  );
end;
/

(end)

Результат выполнения:

(code)

ORA-20150: Ошибка "Internal"!
ORA-06512: at line 16
ORA-20185: Произошла ошибка!
ORA-06512: at line 11

(end)

Пример использования №2:

(code)

declare
  lg lg_logger_t := lg_logger_t.getLogger(
    moduleName => 'Test'
    , objectName => 'TestBlock'
  );

  procedure internal
  is
  begin
    raise_application_error(
      pkg_Error.ProcessError
      , 'Произошла ошибка' || lpad( '!', 1000, '_')
    );
  exception when others then
    raise_application_error(
      pkg_Error.ErrorStackInfo
      , lg.errorStack( 'Ошибка "Internal_"' || lpad( '!', 1000, '_'))
      , true
    );
  end internal;

  procedure internal2
  is
    errorMessage varchar2( 32267);
  begin
    begin
      internal();
    exception when others then
      errorMessage := lg.getErrorStack();
    end;

    raise_application_error(
      pkg_Error.ProcessError
      , 'Произошла ошибка обработки' || lpad( '!', 100, '_')
          || '"' || errorMessage || '"'
    );
  exception when others then
    raise_application_error(
      pkg_Error.ErrorStackInfo
      , lg.errorStack( 'Ошибка "Internal_2"' || lpad( '!', 1000, '_'))
      , true
    );
  end internal2;

begin
  internal2();
exception when others then
  pkg_Common.outputMessage(
    lg.getErrorStack()
  );
end;
/

(end) 

Результат выполнения:

(code)

ORA-20150: Ошибка "Internal2"!
ORA-06512: at line 38
ORA-20185: Произошла ошибка обработки!"ORA-20150: Ошибка "Internal"!
ORA-06512: at line 15
ORA-20185: Произошла ошибка

(end)